opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexDatabase.java

```diff
@@ -1902,54 +1905,61 @@ private void indexParallel(String dir, IndexDownArgs args) {
+                                String errmsg = String.format("ERROR addFile(): '%s'", x.file);
+                                LOGGER.log(Level.SEVERE, errmsg, e);
+                                x.exception = e;
+                                x.ret = false;
+                            } catch (InterruptedException e) {
+                                // Allow one retry if interrupted
+                                if (++tries <= 1) {
+                                    continue;
+                                }
+                                LOGGER.log(Level.WARNING, "No retry: ''{0}''", x.file);
+                                x.exception = e;
+                                x.ret = false;
+                            } catch (RuntimeException | IOException e) {
+                                String errmsg = String.format("ERROR addFile(): '%s'", x.file);
+                                LOGGER.log(Level.WARNING, errmsg, e);
+                                x.exception = e;
+                                x.ret = false;
+                            } finally {
+                                if (pctags != null) {
+                                    pctags.reset();
+                                    ctagsPool.release(pctags);
+                                }
                             }
-                        } catch (AlreadyClosedException e) {
-                            alreadyClosedCounter.incrementAndGet();
-                            String errmsg = String.format("ERROR addFile(): '%s'", x.file);
-                            LOGGER.log(Level.SEVERE, errmsg, e);
-                            x.exception = e;
-                            ret = false;
-                        } catch (InterruptedException e) {
-                            // Allow one retry if interrupted
-                            if (++tries <= 1) {
-                                continue;
-                            }
-                            LOGGER.log(Level.WARNING, "No retry: ''{0}''", x.file);
-                            x.exception = e;
-                            ret = false;
-                        } catch (RuntimeException | IOException e) {
-                            String errmsg = String.format("ERROR addFile(): '%s'", x.file);
-                            LOGGER.log(Level.WARNING, errmsg, e);
-                            x.exception = e;
-                            ret = false;
-                        } finally {
-                            if (pctags != null) {
-                                pctags.reset();
-                                ctagsPool.release(pctags);
-                            }
-                        }
 
-                        progress.increment();
-                        return ret;
-                    }
-                }))).get();
+                            progress.increment();
+                            return x;
```

Comment:
Can we remove this while statement since we are unconditionally returning in first execution of this loop.

Also moving out the entire  lambda expression as a separate method may be easier to read.
Something with signature -> Callable<IndexFileWork> addFileToindex(IndexFileWork x,Map <AtomicInteger> countersMap)

Reply:
The `InterruptedException` handling requires the cycle so that it can retry. The return on the first iteration does not happen in such case.

As for the refactoring by extracting the method, this is something I had done however then reverted because I wanted to keep the changes minimal - it is arguably nicer to observe the changes in the diff view. Might do it later in a subsequent PR.